app/lib/search/index_simple.dart

```diff
@@ -227,29 +230,82 @@ class SimplePackageIndex implements PackageIndex {
 }
 
 class Score {
-  final Map<String, double> values = <String, double>{};
+  final Map<String, double> _values;
+
+  Score(Map<String, double> values) : _values = new Map.unmodifiable(values);
+
+  Set<String> getKeys() => _values.keys.toSet();
+  double getMaxValue() => _values.values.fold(0.0, math.max);
+  Map<String, double> getValues() => _values;
 
-  Iterable<String> getKeys() => values.keys;
+  double operator [](String key) => _values[key] ?? 0.0;
 
-  void addValues(Map<String, double> newValues, double weight) {
-    if (newValues == null) return;
-    newValues.forEach((String key, double score) {
-      if (score != null) {
-        final double prev = values[key] ?? 0.0;
-        values[key] = prev + score * weight;
+  /// Calculates the intersection of the [scores], by multiplying the values.
+  static Score multiply(List<Score> scores) {
+    Set<String> keys;
+    for (Score score in scores) {
+      if (keys == null) {
+        keys = score.getKeys();
+      } else {
+        keys = keys.intersection(score.getKeys());
       }
-    });
+    }
+    return new Score(new Map.fromIterable(
+      keys,
+      value: (key) =>
+          scores.fold(1.0, (double value, Score s) => s[key] * value),
+    ));
+  }
+
+  /// Calculates the union of the [scores], by using the maximum values from
+  /// the sets.
+  static Score max(List<Score> scores) {
+    final result = <String, double>{};
+    for (Score score in scores) {
+      for (String key in score.getKeys()) {
+        result[key] = math.max(result[key] ?? 0.0, score[key]);
+      }
+    }
+    return new Score(result);
   }
 
-  void removeWhere(bool keyCondition(String key)) {
-    final Set<String> keysToRemove = values.keys.where(keyCondition).toSet();
-    keysToRemove.forEach(values.remove);
+  Score removeLowValues({double fraction, double minValue}) {
+    double threshold = minValue;
+    if (fraction != null) {
+      final double fractionValue = getMaxValue() * fraction;
+      threshold ??= fractionValue;
+      threshold = math.max(threshold, fractionValue);
+    }
+    if (threshold == null) {
+      return this;
+    }
+    final result = <String, double>{};
+    for (String key in _values.keys) {
+      final double value = _values[key];
+      if (value < threshold) continue;
+      result[key] = value;
+    }
+    return new Score(result);
   }
 
-  void removeLowScores(double fraction) {
-    final double maxValue = values.values.fold(0.0, max);
-    final double cutoff = maxValue * fraction;
-    removeWhere((key) => values[key] < cutoff);
+  Score project(Iterable<String> keys) {
+    final result = <String, double>{};
+    for (String key in keys) {
+      final double value = _values[key];
+      if (value == null) continue;
+      result[key] = value;
+    }
+    return new Score(result);
+  }
+
+  Score interpolate(double t) {
```

Comment:
How would you call a function that maps the region of [0.0 ... 1.0] values to [0.9 .. 1.0]? Because the role of interpolate is only to do that to prepare multiplication but to prevent to much penalty if e.g. a popularity value is zero.

The problem with the interpolation in your suggestion is that it does seem to depend on the application order of the popularity/health/maintenance values, and therefore not symmetric, not associative. (I haven't actually checked it, but it felt that way). Multiplication on the other hand is associative and doesn't depend on the order, it felt more "right".

Reply:
I have never proposed to use use linear interpolation, I'm just stating the fact that this function is not doing linear interpolation (and thought the previous version of the code did, that's why I suggested the name).

How about `map`:

```dart
final Score result = scoreA.map((double value) => 0.8 + 0.2 * value);
```

?

Reply:
`map` it is then.

Yeah, a previous version was a bit confusing in that regard, because while it appeared from the parameter passing that it does interpolation, it was the same logic as now. It was more memory efficient though, because it spared the creation of the intermediate Map. Nevertheless, I like the current version better, it is easier to follow.